flowdb/bin/build/0030_schema_interactions.sql

```diff
@@ -0,0 +1,159 @@
+  - topup:                      stores the additional topup type data
+-----------------------------------------------------------
+*/
+CREATE SCHEMA IF NOT EXISTS interactions;
+
+    CREATE TABLE IF NOT EXISTS interactions.subscriber(
+
+        id                      BIGSERIAL PRIMARY KEY,
+        msisdn                  TEXT,
+        imei                    TEXT,
+        imsi                    TEXT,
+        tac                     BIGINT REFERENCES infrastructure.tacs(id)
+        );
+
+    CREATE INDEX ON interactions.subscriber (id);
+    CREATE INDEX ON interactions.subscriber (id, msisdn);
+
+
+
+    CREATE TABLE IF NOT EXISTS interactions.locations(
+
+        cell_id                 BIGSERIAL PRIMARY KEY,
+        site_id                 TEXT,
+        mno_cell_code           TEXT
+
+        );
+
+    SELECT AddGeometryColumn('interactions', 'locations', 'position', 4326, 'POINT', 2);
+
+    CREATE INDEX IF NOT EXISTS interactions_locations_position_index
+        ON interactions.locations
+        USING GIST (position);
+
+    CREATE TABLE IF NOT EXISTS interactions.receiving_parties(
+
+        parties_key             BIGSERIAL PRIMARY KEY,
+        cell_id                 BIGINT REFERENCES interactions.locations(cell_id),
+        party_msisdn            TEXT
+
+        );
+
+
+    CREATE SEQUENCE interactions.event_id START 1;
+    
+    CREATE TABLE IF NOT EXISTS interactions.event_supertable (
+        event_id                BIGINT NOT NULL PRIMARY KEY,
+        subscriber_id           BIGINT REFERENCES interactions.subscriber(id),
+        cell_id                 BIGINT REFERENCES interactions.locations(cell_id),
+        time_sk                 BIGINT REFERENCES public.d_time(time_dim_id),
+        date_sk                 BIGINT REFERENCES public.d_date(date_dim_id),
+        event_type              VARCHAR(5) NOT NULL,
+        timestamp               TIMESTAMPTZ NOT NULL
+    );
+
+CREATE TABLE IF NOT EXISTS interactions.calls(
+        event_id                BIGINT NOT NULL DEFAULT nextval('interactions.event_id'),
+        subscriber_id           BIGINT REFERENCES interactions.subscriber(id),
+        cell_id                 BIGINT REFERENCES interactions.locations(cell_id),
+        time_sk                 BIGINT REFERENCES public.d_time(time_dim_id),
+        date_sk                 BIGINT REFERENCES public.d_date(date_dim_id),
+        event_type              VARCHAR(5) NOT NULL DEFAULT 'calls',
+        timestamp               TIMESTAMPTZ NOT NULL,
+        called_subscriber_id    BIGINT REFERENCES interactions.subscriber(id),
+        called_party_cell_id    BIGINT REFERENCES interactions.locations(cell_id),
+        calling_party_msisdn    TEXT,
+        called_party_msisdn     TEXT,
+        call_duration           NUMERIC
+        CHECK ( event_type = 'calls' ), PRIMARY KEY (event_id, date_sk)
+    ) PARTITION BY RANGE (date_sk);
+
+    CREATE TABLE IF NOT EXISTS interactions.sms(
```

Comment:
As I look at this again, I kind of wonder if my attempt to capture the fact than an SMS can go to multiple parties is actually helpful, given that our experience is that the data necessary to allow you to record that accurately has never been available.

Could rejig this a little so that there's an extra level of inheritance, which puts both the calls and sms under an 'interactions' table with their common fields, and treat SMS as only between two parties? Alternatively, would seem reasonable to create a view that captures that information.

Reply:
Would also be possible to:

- Make the supertable a distinct partitioned table
- Make the event type tables separate partitioned tables which ref to the PK of that table and contain only non common info
- Create a view to capture the interactions (calls & sms) commonality

Reply:
I've reworked this so that the events table is a partitioned fact table with an event type field, and additional metadata is split out into the corresponding event type tables.

Also made sms effectively the same as calls, but without a duration which reflects the fact that we've never encountered reliable data to identify multi-recipient sms messages.